homeassistant/components/husqvarna_automower/coordinator.py

```diff
@@ -0,0 +1,44 @@
+"""Data UpdateCoordinator for the Husqvarna Automower integration."""
+import logging
+
+import aioautomower
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_TOKEN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AutomowerDataUpdateCoordinator(DataUpdateCoordinator[None]):
+    """Class to manage fetching Husqvarna data."""
+
+    def __init__(
+        self, hass: HomeAssistant, implementation, access_token, entry: ConfigEntry
+    ) -> None:
+        """Initialize data updater."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+        )
+        self.session = aioautomower.AutomowerSession(
+            implementation.client_id, access_token
+        )
+        self.session.register_token_callback(
```

Comment:
The config entry is updated by our helper when the token is refreshed. We should use that helper interface to refresh the token.

https://github.com/home-assistant/core/blob/b2b57c5f87a519af951a9d1ca8777f5ae3517b92/homeassistant/helpers/config_entry_oauth2_flow.py#L493-L503

Reply:
Right now, the helper function `async_ensure_token_valid()` is not implemented in the integration. Token updates are handled by the library. And the config entry is updated from the library if there is a new token.
What do you mean by "use the helper interface"?

Reply:
The library client should use our helper session to authorize and refresh tokens.

See our library guide and use our scaffold script to get example code. Links here:

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#configuration-via-oauth2

See also eg the Google integration for a real implementation.

Reply:
Should be solved now.